text/0225-configuration.md

```diff
@@ -0,0 +1,329 @@
+
+```python
+
+filepath = "./config.yaml"
+
+try:
+  cfg = opentelemetry.ParseAndValidateConfigurationFromFile(filename)
+except Exception as e:
+  print(e)
+
+filename = "./config.json"
+
+try:
+  cfg = opentelemetry.ParseAndValidateConfigurationFromFile(filename, format="json")
+except Exception as e:
+  raise e
+
+```
+
+#### Go ParseAndValidateConfigurationFromFile example
+
+```go
+
+filename := "./config.yaml"
+cfg, err := otel.ParseAndValidateConfigurationFromFile(filename)
+if err != nil {
+  return err
+}
+
+filename := "./config.json"
+cfg, err := otel.ParseAndValidateConfigurationFromFile(filename, otelconfig.WithFormat("json"))
+if err != nil {
+  return err
+}
+
+```
+
+Implementations *MUST* allow users to specify an environment variable to set the configuration file. This gives flexibility to end users of implementations that do not support command line arguments. The proposed name for this variable:
+
+* `OTEL_CONFIG_FILE`
+
+### Configuration model
+
+To allow SDKs and instrumentation libraries to accept configuration without having to implement the parsing logic, a `Configuration` model *MAY* be provided by implementations. This object:
+
+* has already been parsed from a file or data structure
+* is structurally valid (errors may yet occur when SDK or instrumentation interprets the object)
+
+### Configuration file
+
+The following demonstrates an example of a configuration file format (full example [here](https://github.com/MrAlias/otel-schema/blob/main/config.yaml)):
+
+```yaml
+# include version specification in configuration files to help with parsing and schema evolution.
+file_format: 0.1
+sdk:
+  # Disable the SDK for all signals.
+  #
+  # Boolean value. If "true", a no-op SDK implementation will be used for all telemetry
+  # signals. Any other value or absence of the variable will have no effect and the SDK
+  # will remain enabled. This setting has no effect on propagators configured through
+  # the OTEL_PROPAGATORS variable.
+  #
+  # Environment variable: OTEL_SDK_DISABLED
+  disabled: false
+  # Configure resource attributes and resource detection for all signals.
+  resource:
+    # Key-value pairs to be used as resource attributes.
+    #
+    # Environment variable: OTEL_RESOURCE_ATTRIBUTES
+    attributes:
+      # Sets the value of the `service.name` resource attribute
+      #
+      # Environment variable: OTEL_SERVICE_NAME
+      service.name: !!str "unknown_service"
+  # Configure context propagators. Each propagator has a name and args used to configure it. None of the propagators here have configurable options so args is not demonstrated.
+  #
+  # Environment variable: OTEL_PROPAGATORS
+  propagators: [tracecontext, baggage]
+  # Configure the tracer provider.
+  tracer_provider:
+    # Span exporters. Each exporter key refers to the type of the exporter. Values configure the exporter. Exporters must be associated with a span processor.
+    exporters:
```

Comment:
The goal with this was to reduce nesting. 

Some conversations on this took place [here](https://github.com/MrAlias/otel-schema/pull/16#discussion_r1035353965), and [here](https://github.com/MrAlias/otel-schema/issues/12).

Reply:
Sorry for the slow reply. I was trying to read all the conversation about this before replying. I may be alone here but to me this seems confusing for new users not already very familiar with the limitations of the spec. I think you will find that new users will try to do things like reference the same exporter from multiple processors and will be confused when that breaks. It breaks the actual model of the OTel architecture. "reduce nesting" seems like a technical decision made for aesthetic reasons.

Reply:
If a user has a desire to reduce nesting they can also use YAML anchors

```yaml
exporters:
  - &zipkin
    zipkin:
      endpoint: http://localhost:9411/api/v2/spans
      timeout: 10000

sdk:
  tracer_provider:
    exporter: *otlp-trace
```

Reply:
It's true. I don't feel strongly about this. Interested in what you think about the [sampler config](https://github.com/MrAlias/otel-schema/blob/main/config.yaml#L178-L229) schema we've proposed. Given that samplers are often composed, we've provided:

- `sampler_config`: a map whose keys are the name / type of sampler, and whose values are the config for the sampler. Config is free to reference keys of other samplers in the map to build up higher order samplers. 
- `sampler`: refers to a key in `sampler_config` and dictates which sampler the tracer provider is configured with.

This strategy keeps the sampler configuration at a fixed depth where it might otherwise be arbitrarily deep.

The exporter / span processor strategy is similar to this.

It could be beneficial if we imagine a future where exporters are frequently composed together. This may be a popular thing to do given the lack of ability mutability for `SpanProcessor#onEnd`. No where to filter / obfuscate data except in the exporter layer.

Reply:
I have more or less the same thoughts about the sampler config. In the example you linked I'm not even sure how the Jaeger sampler fits in. It doesn't appear to be reference anywhere? If so, this is exactly the type of configuration mistake/bug that I am worried this type of configuration will cause.

I guess I understand with sampling why you may want to configure a few and compose them together but for exporters that makes a lot less sense to me. I can't think of an example of an exporter that would be composed with another that wouldn't be better solved with a span processor anyway.